File: c++/123_best_time_to_buy_and_sell_stock_iii.cpp
#include <gtest/gtest.h>
#include <iostream>
#include <limits>
#include <vector>

class Solution {
   public:
    int maxProfit(const std::vector<int>& prices) {
        if (prices.empty()) {
            return 0;
        }

        int              n = prices.size();
        std::vector<int> profits(n, 0);

        int maxProfit1 = 0;
        int minPrice   = std::numeric_limits<int>::max();
        for (int i = 0; i < n; i++) {
            minPrice   = std::min(minPrice, prices[i]);
            maxProfit1 = std::max(maxProfit1, prices[i] - minPrice);
            profits[i] += maxProfit1;
        }

        int maxProfit2 = 0;
        int maxPrice   = 0;
        for (int i = n - 1; i > 0; i--) {
            maxPrice   = std::max(maxPrice, prices[i]);
            maxProfit2 = std::max(maxProfit2, maxPrice - prices[i]);
            profits[i - 1] += maxProfit2;
        }

        return *std::max_element(profits.begin(), profits.end());
    }
};

int main(int argc, char* argv[]) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

TEST(test, case1) {
    Solution solution;
    EXPECT_EQ(solution.maxProfit({3, 3, 5, 0, 0, 3, 1, 4}), 6);
    EXPECT_EQ(solution.maxProfit({1, 2, 3, 4, 5}), 4);
    EXPECT_EQ(solution.maxProfit({1}), 0);
    EXPECT_EQ(solution.maxProfit({}), 0);
}

File: python/123_best_time_to_buy_and_sell_stock_iii.py
#!/usr/bin/env python3

import sys


class Solution:
    def maxProfit(self, prices):
        if not prices:
            return 0
        profits = []
        minPrice = sys.maxsize
        profit = 0
        for i in range(0, len(prices)):
            minPrice = min(minPrice, prices[i])
            profit = max(profit, prices[i] - minPrice)
            profits.append(profit)
        maxPrice = 0
        profit = 0
        for i in range(len(prices) - 1, 0, -1):
            maxPrice = max(maxPrice, prices[i])
            profit = max(profit, maxPrice - prices[i])
            profits[i - 1] += profit
        return max(profits)


if __name__ == '__main__':
    print(Solution().maxProfit([7, 1, 5, 3, 6, 4]), 7)
    print(Solution().maxProfit([3, 3, 5, 0, 0, 3, 1, 4]), 6)
    print(Solution().maxProfit([7, 6, 5, 4, 3, 2, 1]), 0)
    print(Solution().maxProfit([1, 2, 3, 4, 5]), 4)
    print(Solution().maxProfit([1]), 0)
    print(Solution().maxProfit([1, 2]), 1)
    print(Solution().maxProfit([1, 2, 3]), 2)
    print(Solution().maxProfit([3, 2, 6, 5, 0, 3]), 7)
